src/modules/commander/Commander.cpp

```diff
@@ -2597,34 +2557,48 @@ Commander::run()
 							send_parachute_command();
 						}
 
-						break;
-					}
-				}
-			}
-
-			_geofence_violated_prev = _geofence_result.geofence_violated;
+					} else if (_geofence_land_on) {
+						// We have a latched geofence land action, do not take any action less severe than terminate
+					} else if (_geofence_result.geofence_action == geofence_result_s::GF_ACTION_LAND) {
+						if (TRANSITION_CHANGED == main_state_transition(_status, commander_state_s::MAIN_STATE_AUTO_LAND, _status_flags,
+								_internal_state)) {
+							_geofence_land_on = true;
+						}
 
-			// reset if no longer in LOITER or if manually switched to LOITER
-			const bool in_loiter_mode = _internal_state.main_state == commander_state_s::MAIN_STATE_AUTO_LOITER;
+					} else if (_geofence_rtl_on) {
+						// We have a latched geofence RTL action, do not take any action less severe than land
+					} else if (_geofence_result.geofence_action == geofence_result_s::GF_ACTION_RTL) {
+						if (TRANSITION_CHANGED == main_state_transition(_status, commander_state_s::MAIN_STATE_AUTO_RTL, _status_flags,
+								_internal_state)) {
+							_geofence_rtl_on = true;
+						}
 
-			if (!in_loiter_mode) {
-				_geofence_loiter_on = false;
-			}
+					} else if (_geofence_loiter_on) {
+						// We have a latched geofence loiter/hold action, do not take any action less severe than land
+					} else if (_geofence_result.geofence_action == geofence_result_s::GF_ACTION_LOITER) {
+						if (TRANSITION_CHANGED == main_state_transition(_status, commander_state_s::MAIN_STATE_AUTO_LOITER, _status_flags,
+								_internal_state)) {
+							_geofence_loiter_on = true;
+						}
 
+					} else if (_geofence_result.geofence_action == geofence_result_s::GF_ACTION_WARN) {
+						// do nothing, mavlink critical messages are sent by navigator
+					} else if (_geofence_result.geofence_action == geofence_result_s::GF_ACTION_NONE) {
+						// do nothing
+					}
 
-			// reset if no longer in RTL or if manually switched to RTL
-			const bool in_rtl_mode = _internal_state.main_state == commander_state_s::MAIN_STATE_AUTO_RTL;
+					_geofence_action_prev = _geofence_result.geofence_action;
+				}
 
-			if (!in_rtl_mode) {
-				_geofence_rtl_on = false;
+			} else {
+				// reset to none when no active breach
+				_geofence_action_prev = geofence_result_s::GF_ACTION_NONE;
```

Comment:
I agree that it would be isolated easier to read this if/else if the shorter block was placed first, but it would also mean we get a less readable "if not <condition> then <cleanup bit>, else <important bit>" rather than more intuitive "if <condition> then <important bit>, else <cleanup bit>"
Also it would move clearing of `_geofence_action_prev` further away from clearing of the related `_geofence_<land|rtl|loiter>_on`